### ISSUE TITLE
Customizable file names pattern

### ISSUE THREAD

**Issue by LeoDupont** [simulot/aspiratv#63]
Hello

It would be nice to be able to set a customized pattern for file names.
For example, with tags like `"{{show}} - {{date}} - {{time}}"`.

_My use case is that the show I'm watching sometimes gets aired twice a day, and I need to be able to differentiate the morning episode from the evening one.
Another reason is that I don't want to keep the show name at the beginning of the file names, for clarity. But for this use case, I could use a custom script after the download._

If you're open to new contributors, I would be happy to give a try at implementing this feature.

**Comment by simulot** [comment]
Well, 

The file name convention is following PLEX, KODI, Jellyfin  and other media servers convention. Read [here](https://support.plex.tv/articles/naming-and-organizing-your-tv-show-files/) for details.

This gives following result directly:
![image](https://user-images.githubusercontent.com/5980377/107532309-0e220180-6bbe-11eb-8a42-a8fd6c0bbf55.png)

Could you give me a concrete example of a show aired twice a day? I'll do some tests.

The duplicate download checker that is currently based naming convention, but I think it's doable to templatize the file name.

Of course you're welcome. I'm struggling  with a potential html frontend for this program. May be we can explore possibilities.

**Comment by LeoDupont** [comment]
> Could you give me a concrete example of a show aired twice a day? I'll do some tests.

Sure, the show I'm talking about is "Tout le monde a son mot à dire", on France 2.
It is usually aired once a day at 18:00, but they often skip the Tuesday evening and air an episode on Friday morning à 10:00.
For example, it has occurred this Friday, February 5th.

The problem is that the two episodes are then renamed the same way, which makes it difficult to differentiate them, and can also cause problems for the duplicate checker, or even while downloading both at the same time.

> The file name convention is following PLEX, KODI, Jellyfin and other media servers convention. Read [here](https://support.plex.tv/articles/naming-and-organizing-your-tv-show-files/) for details.

I understand, and it's a great default. But it might come in handy for other use cases.
Also, I usually don't include the show name in my files and Emby works great without it. But that's just personal preference.

> The duplicate download checker that is currently based naming convention, but I think it's doable to templatize the file name.

I understand, maybe the checker can use the template to detect duplicates.

> Of course you're welcome. I'm struggling with a potential html frontend for this program. May be we can explore possibilities.

Great, I'll start working on it 👍 

Do you want an HTML frontend for entering the pattern?
At first I was thinking of a key in the config file and/or a parameter, like this:
```json
{
	"Watchlist": [
		{
			"Show": "Tout le monde a son mot à dire",
			"Provider": "francetv",
			"Destination": "Emissions",
			"Filename": "{{show}} - {{date}} {{time}} - {{title}}"
		}
	]
}
```

**Comment by simulot** [comment]
> Tout le monde a son mot à dire

For francetv, I grab the list of available  videos on the page "toutes les videos". From this list, I'm building a list of episode to download, and I check if a file is not yet present on the disk. When the file is missing, go on the episode for taking the description, exact air time and media URL.  I'm doing this to limit queries to the server. 

May be we can user the media ID which is expected to be unique, and available on the list. Combined to custom template for file name, it could make it, isn't it? 

> I understand, and it's a great default. But it might come in handy for other use cases.
> Also, I usually don't include the show name in my files and Emby works great without it. But that's just personal preference.

So we have a solid use case. The structure below is holding metadata that can be used for template.  
https://github.com/simulot/aspiratv/blob/7e5f07e3eec3d83b4673a64b00a01cb31a003841/metadata/nfo/nfo.go#L35-L60

Neat, isn't it?

> Great, I'll start working on it +1

We should discuss about this. I have opened the discussion feature  at https://github.com/simulot/aspiratv/discussions

**Comment by simulot** [comment]
Closed with v0.14.0

**Comment by LeoDupont** [comment]
> Closed with v0.14.0

Great work @simulot ! 👍
Thanks a lot for implementing this (and also for the `SeasonPathTemplate`, which is very useful to keep empty for some shows).

There is just one small issue with the aired time.
If I use a template with date and time aired, like this: `{{.Aired.Time.Format "2006-01-02 Monday 15h04"}}`, only the .nfo gets the time while the .mp4 gets 00h00.

Example (took a random show with short episodes so the download is quick):

Watchlist:
```json
{
	"Show": "Parole inattendue",
	"Provider": "francetv",
	"Destination": "Videos",
	"SeasonPathTemplate": "",
	"ShowNameTemplate": "{{.Showtitle}} - {{.Aired.Time.Format \"2006-01-02 Monday 15h04\"}} - {{.Title}}.mp4"
}
```
Downloading:

![image](https://user-images.githubusercontent.com/29914053/107895474-c12e8a00-6f33-11eb-9b9d-6f1d5b4a3c39.png)

End result:

![image](https://user-images.githubusercontent.com/29914053/107895479-c4c21100-6f33-11eb-86d9-600f705b6aeb.png)

**Comment by simulot** [comment]
I haven't tested with hour part in the format. Interesting effect. 

The index page gives the full list of available shows, but only broadcast date is available:
![image](https://user-images.githubusercontent.com/5980377/107916504-ae817880-6f66-11eb-949b-5ff458b5ee7a.png)

This gives the name the name for the mp1 file. If the file doesn't exist, detailed information is pulled from show page that give the the time as well and the plot, and information to get media's url.
![image](https://user-images.githubusercontent.com/5980377/107917102-aece4380-6f67-11eb-9da9-c10f3832a5ad.png)

The episode page must be fetched to get the exact time. Most of the time, this is unnecessary, because we have already the episode. I would like to keep lowest possible profile and limiting the noise we make. The idea is to not disturb and stay tolerated. 

A better technical solution would be keeping track of downloaded episodes with unique IDs. Episode's ID would be checked against this list and download it when not present. I'm thinking about it for managing a front end. 

But for now, I'd prefer step back, and not getting exact time.
